src/lint/track_config.nim

```diff
@@ -621,6 +621,39 @@ proc satisfiesSecondPass(trackConfigContents: string; path: Path): bool =
   checkExercisesPCP(practiceExercises, result, path)
   checkExerciseSlugsAndForegone(exercises, result, path)
 
+proc getExerciseSlugs(data: JsonNode; k: string): HashSet[string] =
+  result = initHashSet[string]()
+  if data.kind == JObject:
+    if data.hasKey("exercises"):
+      let exercises = data["exercises"]
+      if exercises.kind == JObject:
+        if exercises.hasKey(k):
+          if exercises[k].kind == JArray:
+            for exercise in exercises[k]:
+              if exercise.kind == JObject:
+                if exercise.hasKey("slug"):
+                  let slug = exercise["slug"]
+                  if slug.kind == JString:
+                    let slugStr = slug.getStr()
+                    if slugStr.len > 0:
+                      result.incl slugStr
```

Comment:
So.... I take you're not a fan of early returns? :)

Reply:
Mmm, it depends. This was a silly block of code, but here I think it's ugly regardless of how it's written.

I've changed to the below:
```Nim
proc getExerciseSlugs(data: JsonNode; k: string): HashSet[string] =
  result = initHashSet[string]()
  if data.kind != JObject or "exercises" notin data:
    return

  let exercises = data["exercises"]

  if exercises.kind != JObject or k notin exercises or exercises[k].kind != JArray:
    return

  for exercise in exercises[k]:
    if exercise.kind == JObject:
      if exercise.hasKey("slug"):
        let slug = exercise["slug"]
        if slug.kind == JString:
          let slugStr = slug.getStr()
          if slugStr.len > 0:
            result.incl slugStr
```

But I claim that it isn't a strict improvement. The indentation reflected the underlying design issue, and early `return` conceals it, making it less likely to be refactored in the future :)

Background: I wanted to check the directories even if the exercise part of the `config.json` is invalid - so with the current deserialization design, and without doing some refactoring, or a big `try`, we're stuck with validating the types a second time.


The conditions would be much more readable using some helper procs named e.g. `hasString`, but those procs we've implemented also write an error message if the condition is false. And I didn't want to print some duplicate error messages, or refactor those procs to take an extra parameter, so...